modules/4_nodes/nodes.tf

```diff
@@ -260,3 +260,19 @@ data "ibm_pi_instance_ip" "worker_ip" {
   pi_network_name      = var.network_name
   pi_cloud_instance_id = var.service_instance_id
 }
+
+data "ibm_pi_instance" "master_id" {
+  depends_on = [ibm_pi_instance.master]
+  count      = var.master["count"]
+
+  pi_instance_name     = ibm_pi_instance.master[count.index].pi_instance_name
+  pi_cloud_instance_id = var.service_instance_id
+}
+
+data "ibm_pi_instance" "worker_id" {
+  depends_on = [ibm_pi_instance.worker]
+  count      = var.worker["count"]
+
+  pi_instance_name     = ibm_pi_instance.worker[count.index].pi_instance_name
+  pi_cloud_instance_id = var.service_instance_id
+}
```

Comment:
Why introduce data sources? You may get the list of instance_id using the resource attribute references.

Reply:
@yussufsh did try that but then `terraform apply` on 2nd attempt was in case of failure on first was destroying and recreating nodes - so decided to opt for data source approach. @pravin-dsilva (any other observations you made in your very first try). I can still give it a try using approach without data sources and get back to you.

Reply:
Fixed.